File: backend/app/routes/applications.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import List, Optional

from app.database import get_db
from app.models.application import Application
from app.schemas.application import ApplicationCreate, ApplicationUpdate, ApplicationResponse

router = APIRouter(prefix="/api/applications", tags=["applications"])


@router.post("/", response_model=ApplicationResponse, status_code=201)
async def create_application(payload: ApplicationCreate, db: AsyncSession = Depends(get_db)):
    application = Application(**payload.model_dump())
    db.add(application)
    await db.commit()
    await db.refresh(application)
    return application


@router.get("/", response_model=List[ApplicationResponse])
async def list_applications(status: Optional[str] = None, db: AsyncSession = Depends(get_db)):
    query = select(Application)
    if status:
        query = query.where(Application.status == status)
    result = await db.execute(query)
    return result.scalars().all()


@router.get("/{application_id}", response_model=ApplicationResponse)
async def get_application(application_id: int, db: AsyncSession = Depends(get_db)):
    application = await db.get(Application, application_id)
    if not application:
        raise HTTPException(status_code=404, detail="Application not found")
    return application


@router.patch("/{application_id}", response_model=ApplicationResponse)
async def update_application(
    application_id: int, payload: ApplicationUpdate, db: AsyncSession = Depends(get_db)
):
    application = await db.get(Application, application_id)
    if not application:
        raise HTTPException(status_code=404, detail="Application not found")

    update_data = payload.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(application, field, value)

    await db.commit()
    await db.refresh(application)
    return application


@router.delete("/{application_id}", status_code=204)
async def delete_application(application_id: int, db: AsyncSession = Depends(get_db)):
    application = await db.get(Application, application_id)
    if not application:
        raise HTTPException(status_code=404, detail="Application not found")
    await db.delete(application)
    await db.commit()

File: frontend/src/api/applications.ts
import { apiClient } from "./client";
import type { Application, ApplicationCreate } from "../types/application";

export async function getApplications(status?: string): Promise<Application[]> {
  const response = await apiClient.get<Application[]>("/api/applications/", {
    params: status ? { status } : {},
  });
  return response.data;
}

export async function getApplication(id: number): Promise<Application> {
  const response = await apiClient.get<Application>(`/api/applications/${id}`);
  return response.data;
}

export async function createApplication(data: ApplicationCreate): Promise<Application> {
  const response = await apiClient.post<Application>("/api/applications/", data);
  return response.data;
}

export async function updateApplication(
  id: number,
  data: Partial<ApplicationCreate>
): Promise<Application> {
  const response = await apiClient.patch<Application>(`/api/applications/${id}`, data);
  return response.data;
}

export async function deleteApplication(id: number): Promise<void> {
  await apiClient.delete(`/api/applications/${id}`);
}